### ISSUE TITLE
Re-work self-test code.

### ISSUE THREAD

**Issue by magnumripper** [openwall/john#948]
The current self-test code differs from IRL runs. For example, it will do things like this:
1. call clear_keys()
2. call set_key(0)
3. call crypt_all(1)
4. call set_key(1)
5. call crypt_all(2)

We should re-work it but we need to think it over very carefully.

**Comment by jfoug** [comment]
Is this what @loverszhaokai is working on?

**Comment by magnumripper** [comment]
Not really, but his work might obsolete this. I haven't even tested his code yet.

**Comment by loverszhaokai** [comment]
@jfoug  I am not working on this. What I have done is detecting FMT_* flags and simulate the real crypt_all() which call max times of set_key().
